Feat/#220/caching data objects

kadmit opened the pull request:
Task: https://github.com/keeleinstituut/tv-tolkevarav/issues/220

The PR contains endpoints for the synchronization of classifier values with other services.

Commit:
feat(220) - add observer && publishing of classifier events to the queue.

Commit:
feat(220) - dependency update

Commit:
feat(220) - dependency update && add endpoints for sync of classifier values with other services && pint style fixes

Commit:
feat(220) - Fix of rate limiting for sync endpoints

Commit:
feat(220) - Pint style fixes

Commit:
feat(220) - Update dependencies

Commit:
feat(220) - Add tests for sync endpoints

Commit:
feat(220) - Fix phpunit.xml

Commit:
feat(220) - pint style fixes

Commit:
feat(220) - dependencies update && fix tests

DanBondarenko commented on application/app/Observers/ClassifierValueObserver.php:
```diff
@@ -0,0 +1,48 @@
+        $this->publishEvent($classifierValue, 'classifier-value.deleted');
+    }
+
+    private function publishEvent(ClassifierValue $classifierValue, string $routingKey = ''): void
```
**[non-blocker]** Hmm, is it ever appropriate to leave the routing key unspecified? It seems to me that it is always an error. Therefore I would consider making it a required parameter.

```suggestion
    private function publishEvent(ClassifierValue $classifierValue, string $routingKey): void
```

DanBondarenko commented on application/app/Http/Resources/ClassifierValueSyncResource.php:
```diff
@@ -0,0 +1,29 @@
+            'meta' => $this->getMetaData(),
+            'deleted_at' => $this->deleted_at,
+        ];
+    }
```
**[non-blocker]** Is there a good reason for not including timestamps (`created_at`, `updated_at`). It seems to me that we might as well include all data when syncing. But maybe there’s a good reason why not to do it.

If all data is included, then one approach could be:

```suggestion
    public function toArray(Request $request): array
    {
        return [
            ...$this->resource->toArray(),
            'meta' => $this->getMetaData()
        ];
    }
```

thenouan replied:
My thoughts on this and not a direct recommendation, just how I'm thinking of the synchronization between services.

We can provide all of it as well, but I would ask myself "are these fields necessary?". For example I can't think of any usage for classifier's created_at or updated_at fields on the consumer services side (e.g. translation-order-service relying on classifier values to perform validation and increase data integrity). I would include only necessary data to reduce load on background communication between the services. And we don't always have to expose the models directly, we can transform them to some other structure which would be easier to adopt on the consumer service side, which would yield to decreased amount of requests and size of data

DanBondarenko commented on docker-compose.yml:
```diff
@@ -4,22 +4,12 @@ services:
       - POSTGRES_USER=user
-      - POSTGRES_PASSWORD=pass
-
-  rabbitmq:
-    image: rabbitmq:latest
-    ports:
-      - "5672:5672"
-      - "15672:15672"
-    environment:
-      - RABBITMQ_DEFAULT_USER=guest
-      - RABBITMQ_DEFAULT_PASS=guest
```
**[blocker]** Is there a good reason why the `rabbitmq` service was removed? The tests fail if there’s no connection to RabbitMQ. If `docker compose` is not the intended way to test, then I think there should be some directions in the `README.md` on how to test/run the application.

thenouan replied:
I think the reason was that there is one rabbitmq instance for all of the services, not that every service has its own. So I'm guessing Dima moved it to some more generic place

DanBondarenko replied:
OK, I understand that. But I think since it’s needed for tests to pass, I would personally keep it in the compose project.

DanBondarenko commented on application/routes/api.php:
```diff
@@ -14,5 +15,10 @@
+Route::withoutMiddleware(['auth:api', 'throttle:api'])->group(function () {
+    Route::get('/sync/classifier-values', [ClassifierValueSyncController::class, 'index']);
+    Route::get('/sync/classifier-values/{id}', [ClassifierValueSyncController::class, 'show'])->whereUuid('id');
+});
```
**[non-blocker]** It appears to me that currently the `"auth"` middleware is not being applied to any route in this application (see `RouteServiceProvider.php`). Therefore it’s redundant to exclude it (and might mislead someone to think that it’s included for other routes).

```suggestion
Route::withoutMiddleware('throttle:api')->group(function () {
    Route::get('/sync/classifier-values', [ClassifierValueSyncController::class, 'index']);
    Route::get('/sync/classifier-values/{id}', [ClassifierValueSyncController::class, 'show'])->whereUuid('id');
});
```

thenouan replied:
I think it would be good to have auth middleware applied globally and endpoints that should be publicly accessible will be using the withoutmiddleware. The reason for it is that the whole system in general requires authentication to access it and there's currently no functionalities that don't require authentication except the login page. This would also decrease the risk that some endpoint will be left unguarded

DanBondarenko commented on application/app/Observers/ClassifierValueObserver.php:
**[non-blocker]** It seems to me that there’s no tests which validate that this observer does what is intended. Ideally tests would confirm that these messages are actually being published etc. But I also understand it might not be a priority time-wise. I’ll leave the decision to you.

DanBondarenko commented on docker-compose.yml:
```diff
@@ -4,22 +4,12 @@ services:
-      - RABBITMQ_DEFAULT_USER=guest
-      - RABBITMQ_DEFAULT_PASS=guest
\ No newline at end of file
+      - POSTGRES_PASSWORD=pass
```
**[blocker]** It appears that this compose file can’t be used "out-of-the-box" for testing because `phpunit.xml` expects a database with name "testing", but no such database is created here. 

Other repositories have solved this by creating a `dev-db-init.sql` file and mounting it in the Postgres container (see [tv-authorization’s compose file](https://github.com/keeleinstituut/tv-authorization/blob/main/docker-compose.yml)).

DanBondarenko reviewed (changes requested):
Thank you for the PR! Left some comments. :)

Commit:
feat(220) - remove redundant exclusion of 'auth:api' middleware

Commit:
feat(220) - Add setup of databases using docker

DanBondarenko commented:
Hi @kadmit, thank you for the changes! I would personally re-add the `rabbitmq` to `docker-compose.yml` so that test results would be unambigous, but I don’t want to block progress in case you have a different opinion, so approving the request. :)

Commit:
feat(220) - Add auth for service endpoints && add missing docker compose config && refactor according to the common codebase approach && update dependencies.

kadmit commented:
@DanBondarenko @thenouan  Thank you for the review! I added the missing Rabbitmq configuration to docker-compose. I also added auth for classifier value endpoints according to the [comment](https://github.com/keeleinstituut/tv-classifier/pull/8#discussion_r1222509323) and did minor refactoring according to our codebase style approach 🙂

Commit:
feat(#220) - Add missing params into .env.example && remove unused params from phpunit.xml

Commit:
Merge branch 'main' into feat/#220/caching-data-objects

# Conflicts:
#	docker-compose.development.yml
